README.md
# tdam_api
Python client for the TDAmeritrade API


CHANGELOG.rst

Changelog
=========

0.1.0 (2019-08-22)
------------------

* First release on PyPI.


tests/test_auth.py
from requests import HTTPError
import pytest
from unittest import mock
import responses

from tdam_api import TDClient
from tdam_api.urls import Urls
from tdam_api.entities import AuthenticationRequired, Quote


def test_auth_required():
    c = TDClient(authenticated=False)

    with pytest.raises(AuthenticationRequired):
        c._update_access_token()


def test_missing_credentials():
    with mock.patch.dict("os.environ", clear=True):
        with pytest.raises(Exception) as exc_info:
            c = TDClient()
            assert c is None
        assert "Missing Credentials" in str(exc_info.value)


@responses.activate
def test_quote_auth():
    c = TDClient(access_token="Invalid", refresh_token="Invalid", app_id="Invalid")
    responses.add(
        responses.GET,
        Urls.quote + f"?symbol=FB",
        json={"FB": {"symbol": "FB"}},
        status=200,
    )

    res = c.quote("FB")
    assert isinstance(res, Quote)
    assert res.symbol == "FB"


@responses.activate
def test_access_token_refresh_fail():
    c = TDClient(access_token="Invalid", refresh_token="Invalid", app_id="Invalid")
    responses.add(responses.GET, Urls.quote + f"?symbol=FB", json={}, status=401)

    with mock.patch.object(TDClient, "_update_access_token") as m:
        with pytest.raises(HTTPError):
            c.quote("FB")
        m.assert_called_once()


@responses.activate
def test_access_token_refresh_success():
    c = TDClient(access_token="Invalid", refresh_token="Invalid", app_id="Invalid")
    responses.add(responses.GET, Urls.quote + f"?symbol=FB", json={}, status=401)
    responses.add(
        responses.GET,
        Urls.quote + f"?symbol=FB",
        json={"FB": {"symbol": "FB"}},
        status=200,
    )

    with mock.patch.object(TDClient, "_update_access_token") as m:
        res = c.quote("FB")
        assert isinstance(res, Quote)
        assert res.symbol == "FB"
        m.assert_called_once()


@pytest.mark.local
def test_auth_refresh():
    c = TDClient(access_token="Invalid token")
    with mock.patch.object(
        TDClient, "_update_access_token", side_effect=c._update_access_token
    ) as m:
        res = c.quote("AAPL")
        m.assert_called_once()
        assert isinstance(res, Quote)
        assert res.bidPrice is not None
        assert res.symbol == "AAPL"
        assert res.exchangeName == "NASD"

    with mock.patch.object(
        TDClient, "_update_access_token", side_effect=c._update_access_token
    ) as m:
        res = c.quote("FB")
        m.assert_not_called()
        assert isinstance(res, Quote)
        assert res.bidPrice is not None
        assert res.symbol == "FB"
        assert res.exchangeName == "NASD"


tests/test_client.py
import os
from datetime import datetime, timedelta

import pytest
import responses
from unittest import mock

from tdam_api import TDClient
from tdam_api.entities import (
    SymbolNotFound,
    Quote,
    Stock,
    Instrument,
    Fundamental,
    InvalidArgument,
)
from tdam_api.urls import Urls

apikey = os.environ["TDAM_APP_ID"]


def quote_resp(symbol: str, exchange: str = "NASDAQ") -> dict:
    out = {f"{symbol}": {"symbol": f"{symbol}", "exchange": f"{exchange}"}}
    return out


def fundamental_resp(symbol: str):
    out = {
        f"{symbol}": {
            "fundamental": {
                "symbol": f"{symbol}",
                "marketCap": 234343,
                "peRatio": 18.15,
            }
        }
    }
    return out


def validate_quote_response(res: Quote, symbol: str):
    assert res.bidPrice is not None
    assert res.symbol == symbol
    assert res.exchangeName == "NASD"


@responses.activate
def test_quote():
    c = TDClient(authenticated=False)
    sym = "FB"

    responses.add(
        responses.GET,
        Urls.quote + f"?apikey={apikey}&symbol={sym}",
        json=quote_resp(sym),
        status=200,
    )
    res = c.quote(sym)
    assert res.symbol == sym

    # Test the automatic upper casing
    res = c.quote("fb")
    assert res.symbol == sym

    responses.add(
        responses.GET,
        Urls.quote + f"?apikey={apikey}&symbol=NO DICE",
        json={},
        status=200,
    )
    with pytest.raises(SymbolNotFound):
        c.quote("No Dice")


@responses.activate
def test_find_instrument():
    c = TDClient(authenticated=False)
    # TD API accepts lowercase for the search endpoint
    sym = "lyf.*"

    responses.add(
        responses.GET,
        Urls.search + f"?apikey={apikey}&symbol={sym}&projection=symbol-regex",
        json=quote_resp("LYFT"),
        status=200,
    )
    res = c.find_instrument(sym)
    assert len(res) == 1
    assert isinstance(res["LYFT"], Instrument)


@responses.activate
def test_get_fundamentals():
    c = TDClient(authenticated=False)
    sym = "AAPL"

    responses.add(
        responses.GET,
        Urls.search + f"?apikey={apikey}&symbol={sym}&projection=fundamental",
        json=fundamental_resp(sym),
        status=200,
    )
    res = c.get_fundamentals(sym.lower())
    assert isinstance(res, Fundamental)
    assert res.symbol == sym
    assert res.peRatio == 18.15


def test_stock():
    c = TDClient(authenticated=False)
    with mock.patch.object(TDClient, "quote") as m:
        c.stock("fb")
        m.assert_called_once()
        m.assert_called_with("fb")


def test_get_history():
    c = TDClient(authenticated=False)
    with pytest.raises(InvalidArgument):
        c.get_history(
            "aapl", start_dt=datetime(2019, 1, 31), end_dt=datetime(2019, 1, 1)
        )

    with pytest.raises(InvalidArgument):
        c.get_history("aapl")

    with pytest.raises(InvalidArgument):
        c.get_history(
            "aapl",
            start_dt=datetime(2019, 1, 1),
            end_dt=datetime(2019, 1, 31),
            freq="5d",
        )

    # Ensure that we don't need to pass a end_dt (defaults to datetime.today())
    with mock.patch.object(TDClient, "_get_with_retry") as m:
        m.return_value.json.return_value = {"empty": False, "candles": ["1", "2"]}
        out = c.get_history("aapl", start_dt=datetime(2019, 1, 1))
        m.assert_called_once()
        assert out == ["1", "2"]


@pytest.mark.apitest
def test_quote_unauth():
    c = TDClient(authenticated=False)

    # Test Symbol not found
    with pytest.raises(SymbolNotFound):
        c.quote("No Dice")

    # Test valid case
    res = c.quote("fb")
    assert isinstance(res, Quote)
    validate_quote_response(res, "FB")

    res = c.stock("fb")
    assert isinstance(res, Stock)
    validate_quote_response(res, "FB")


@pytest.mark.apitest
def test_quotes_unauth():
    c = TDClient(authenticated=False)

    # Test Symbol not found
    with pytest.raises(SymbolNotFound):
        c.quotes(["No Dice", "None Here Too"])

    # Test valid case
    res = c.quotes(["fb", "msft"])
    assert len(res.keys()) == 2
    assert "FB" in res.keys()
    assert "MSFT" in res.keys()
    validate_quote_response(res["FB"], "FB")
    validate_quote_response(res["MSFT"], "MSFT")


@pytest.mark.apitest
def test_find_instrument_unauth():
    c = TDClient(authenticated=False)
    res = c.find_instrument("lyft.*")
    # Expecting LYFT
    assert len(res.keys()) == 1
    assert "LYFT" in res.keys()
    assert isinstance(res["LYFT"], Instrument)
    assert res["LYFT"].symbol == "LYFT"
    assert res["LYFT"].exchange == "NASDAQ"


@pytest.mark.apitest
def test_get_fundamentals_unauth():
    c = TDClient(authenticated=False)
    res = c.get_fundamentals("aapl")
    assert isinstance(res, Fundamental)
    assert res.symbol == "AAPL"
    assert res.peRatio is not None
    assert res.marketCap is not None


@pytest.mark.apitest
def test_get_history_unauth():
    c = TDClient(authenticated=False)
    res = c.get_history(
        "aapl", start_dt=datetime(2019, 1, 1), end_dt=datetime(2019, 1, 31)
    )
    assert len(res) == 21
    assert len(res[0].keys()) == 6
    keys = ["open", "high", "low", "close", "volume", "datetime"]
    for k in keys:
        assert k in res[0].keys()

    res = c.get_history(
        "No Dice", start_dt=datetime(2019, 1, 1), end_dt=datetime(2019, 1, 31)
    )
    assert res is None


@pytest.mark.apitest
def test_get_history_df_unauth():
    import pandas as pd

    c = TDClient(authenticated=False)
    res = c.get_history_df(
        "aapl", start_dt=datetime(2019, 1, 1), end_dt=datetime(2019, 1, 31)
    )
    assert isinstance(res, pd.DataFrame)
    assert res.shape == (21, 5)
    assert res.index.name == "datetime"
    keys = ["open", "high", "low", "close", "volume"]
    for k in res.columns:
        assert k in keys

    res = c.get_history_df(
        "No Dice", start_dt=datetime(2019, 1, 1), end_dt=datetime(2019, 1, 31)
    )
    assert res is None


@pytest.mark.apitest
def test_get_intraday_history_df_unauth():
    import pandas as pd

    c = TDClient(authenticated=False)
    res = c.get_history_df(
        "aapl",
        start_dt=datetime.today() - timedelta(4),
        end_dt=datetime.today(),
        freq="30min",
    )
    assert isinstance(res, pd.DataFrame)
    assert res.shape[1] == 5
    assert res.index.name == "datetime"
    keys = ["open", "high", "low", "close", "volume"]
    for k in res.columns:
        assert k in keys

    with pytest.raises(InvalidArgument):
        res = c.get_history_df(
            "AAPL",
            start_dt=datetime.today() - timedelta(60),
            end_dt=datetime.today() - timedelta(55),
            freq="30min",
        )


tests/test_options.py
import os
import json

import pytest
import responses

from tdam_api import TDClient
from tdam_api.entities import (
    InvalidArgument,
    SymbolNotFound,
    Option,
    OptionChain,
    VerticalSpread,
    Straddle,
    Strangle,
)
from tdam_api.urls import Urls

apikey = os.environ["TDAM_APP_ID"]


@responses.activate
def test_get_expirations():
    c = TDClient(authenticated=False)
    sym = "AAPL"

    with open("tests/data/aapl_all_exp_calls.json", "r") as json_file:
        chain_resp = json.load(json_file)

    responses.add(
        responses.GET,
        Urls.option_chain
        + (
            f"?apikey={apikey}&symbol={sym}"
            f"&contractType=CALL&strategy=SINGLE&range=NTM"
        ),
        json=chain_resp,
        status=200,
    )

    expiries = c.get_expirations(sym)
    expected = [
        "2019-08-23",
        "2019-08-30",
        "2019-09-06",
        "2019-09-13",
        "2019-09-20",
        "2019-09-27",
        "2019-10-18",
        "2019-11-15",
        "2019-12-20",
        "2020-01-17",
        "2020-03-20",
        "2020-04-17",
        "2020-06-19",
        "2020-09-18",
        "2021-01-15",
        "2021-06-18",
    ]

    assert sorted(expiries) == sorted(expected)


@responses.activate
def test_get_option_chain():
    c = TDClient(authenticated=False)
    sym = "AAPL"
    expiry = "2019-08-23"

    with pytest.raises(InvalidArgument):
        c.get_option_chain()
    with pytest.raises(InvalidArgument):
        c.get_option_chain(sym)
    with pytest.raises(InvalidArgument):
        c.get_option_chain(expiry=expiry)

    with open("tests/data/aapl_one_expiry.json", "r") as json_file:
        chain_resp = json.load(json_file)

    responses.add(
        responses.GET,
        Urls.option_chain
        + (
            f"?apikey={apikey}&symbol={sym}"
            f"&strategy=SINGLE&fromDate={expiry}&toDate={expiry}"
        ),
        json=chain_resp,
        status=200,
    )
    chain = c.get_option_chain(sym, expiry)
    expected_strikes = list(range(130, 150, 5))
    expected_strikes.extend([x / 10 for x in range(1500, 2501, 25)])

    assert isinstance(chain, OptionChain)
    assert len(chain._calls) == len(expected_strikes)
    assert len(chain._puts) == len(expected_strikes)
    for strike in expected_strikes:
        assert isinstance(chain.get(strike, "C"), Option)
        assert isinstance(chain.get(strike, "P"), Option)

    with pytest.raises(InvalidArgument):
        chain.get(200, "")
    with pytest.raises(InvalidArgument):
        chain.get(200, "call_option")

    vs = chain.get_vertical("C", 145, 150)
    assert isinstance(vs, VerticalSpread)
    assert isinstance(vs.long_option, Option)
    assert isinstance(vs.short_option, Option)
    assert vs.long_option.strikePrice == 145.0
    assert vs.short_option.strikePrice == 150.0

    s = chain.get_straddle(200)
    assert isinstance(s, Straddle)
    assert isinstance(s.call_option, Option)
    assert isinstance(s.put_option, Option)
    assert s.call_option.strikePrice == 200.0
    assert s.put_option.strikePrice == 200.0

    s = chain.get_strangle(200, 202.5)
    assert isinstance(s, Strangle)
    assert isinstance(s.call_option, Option)
    assert isinstance(s.put_option, Option)
    assert s.call_option.strikePrice == 200.0
    assert s.put_option.strikePrice == 202.5


@responses.activate
def test_get_option():
    c = TDClient(authenticated=False)
    sym = "AAPL"
    expiry = "2019-08-23"
    strike = 200

    with open("tests/data/aapl_200.json", "r") as json_file:
        chain_resp = json.load(json_file)

    responses.add(
        responses.GET,
        Urls.option_chain
        + (
            f"?apikey={apikey}&symbol={sym}"
            f"&contractType=CALL&strategy=SINGLE&strike=200"
            f"&fromDate={expiry}&toDate={expiry}"
        ),
        json=chain_resp,
        status=200,
    )

    opt = c.get_option(symbol=sym, expiry=expiry, right="C", strike=strike)

    assert isinstance(opt, Option)
    assert opt.strikePrice == 200.0
    assert opt.putCall == "CALL"
    assert opt.lastTradingDay == 1566532800000
    assert sym in opt.symbol

    responses.add(
        responses.GET,
        Urls.option_chain
        + (
            f"?apikey={apikey}&symbol={sym}"
            f"&contractType=PUT&strategy=SINGLE&strike=200"
            f"&fromDate={expiry}&toDate={expiry}"
        ),
        json=chain_resp,
        status=200,
    )

    opt = c.get_option(symbol=sym, expiry=expiry, right="P", strike=strike)

    assert isinstance(opt, Option)
    assert opt.strikePrice == 200.0
    assert opt.putCall == "PUT"
    assert opt.lastTradingDay == 1566532800000
    assert sym in opt.symbol

    with open("tests/data/no_dice.json", "r") as json_file:
        chain_resp = json.load(json_file)

    sym = "NO DICE"
    responses.add(
        responses.GET,
        Urls.option_chain
        + (
            f"?apikey={apikey}&symbol={sym}"
            f"&contractType=CALL&strategy=SINGLE&strike=200"
            f"&fromDate={expiry}&toDate={expiry}"
        ),
        json=chain_resp,
        status=200,
    )

    with pytest.raises(SymbolNotFound):
        c.get_option(symbol="NO DICE", expiry=expiry, right="C", strike=strike)


src/tdam_api/entities.py
from typing import List, Dict, Any

import attr


@attr.s(frozen=True)
class Entity:
    _data: Dict[str, Any] = attr.ib()

    def __getattr__(self, name: str) -> Any:
        if name in self._data:
            return self._data[name]
        return self.__getattribute__(name)

    def keys(self) -> List[str]:
        return self._data.keys()

    def __getitem__(self, key):
        return self._data[key]

    def _get_data(self):
        return self._data


# For asset based features to be added later
class Quote(Entity):
    pass


class Instrument(Entity):
    pass


class Fundamental(Entity):
    pass


class Stock(Entity):
    pass


class Option(Entity):
    @classmethod
    def float_to_strike(cls, strike: float = None) -> str:
        return str(round(strike + 0.0, 4))


@attr.s(frozen=True)
class VerticalSpread:
    long_option: Option = attr.ib(validator=attr.validators.instance_of(Option))
    short_option: Option = attr.ib(validator=attr.validators.instance_of(Option))


@attr.s(frozen=True)
class Straddle:
    call_option: Option = attr.ib(validator=attr.validators.instance_of(Option))
    put_option: Option = attr.ib(validator=attr.validators.instance_of(Option))


class Strangle(Straddle):
    pass


@attr.s(frozen=True)
class OptionChain:
    _calls: Dict[str, Option] = attr.ib(validator=attr.validators.instance_of(dict))
    _puts: Dict[str, Option] = attr.ib(validator=attr.validators.instance_of(dict))

    def get(self, strike: float, right: str) -> Option:
        key = Option.float_to_strike(strike)
        if right.lower() == "c" or right.lower() == "call":
            return self._calls.get(key, None)
        if right.lower() == "p" or right.lower() == "put":
            return self._puts.get(key, None)
        raise InvalidArgument("right should be one of (c)all or (p)ut")

    def get_vertical(
        self, right: str = "C", long_strike: float = None, short_strike: float = None
    ) -> VerticalSpread:
        long_opt = self.get(long_strike, right)
        short_opt = self.get(short_strike, right)
        return VerticalSpread(long_opt, short_opt)

    def get_straddle(self, strike: float = None) -> Straddle:
        call_opt = self.get(strike, "C")
        put_opt = self.get(strike, "P")
        return Straddle(call_opt, put_opt)

    def get_strangle(
        self, call_strike: float = None, put_strike: float = None
    ) -> Strangle:
        call_opt = self.get(call_strike, "C")
        put_opt = self.get(put_strike, "P")
        return Strangle(call_opt, put_opt)


class Order(Entity):
    pass


# Custom Exceptions
class SymbolNotFound(ValueError):
    pass


class InvalidArgument(ValueError):
    pass


class AuthenticationRequired(Exception):
    pass


src/tdam_api/client.py
import os
import functools
from typing import List, Dict
from datetime import datetime, timedelta

import requests

from .urls import Urls
from .entities import (
    Quote,
    Instrument,
    Fundamental,
    Stock,
    Option,
    OptionChain,
    SymbolNotFound,
    InvalidArgument,
    AuthenticationRequired,
)


def auth_required(f):
    @functools.wraps(f)
    def wrapper(*args, **kwargs):
        if args[0]._authenticated:
            return f(*args, **kwargs)
        else:
            raise AuthenticationRequired("Method requires authentication")

    return wrapper


class TDClient:
    def __init__(
        self, access_token=None, refresh_token=None, app_id=None, authenticated=True
    ):
        if authenticated:
            self.access_token = self._get_auth_var(access_token, "TDAM_ACCESS_TOKEN")
            self.refresh_token = self._get_auth_var(refresh_token, "TDAM_REFRESH_TOKEN")

        self.app_id = self._get_auth_var(app_id, "TDAM_APP_ID")
        self._authenticated = authenticated

    def _get_auth_var(self, param: str, env_var: str) -> str:
        if param is None:
            if env_var in os.environ:
                return os.environ[env_var]
            else:
                raise Exception("Missing Credentials.")
        else:
            return param

    @auth_required
    def _update_access_token(self):
        data = {
            "grant_type": "refresh_token",
            "refresh_token": self.refresh_token,
            "client_id": self.app_id,
        }
        resp: requests.Response = requests.post(Urls.auth, data=data)
        if resp.status_code == 200:
            self.access_token = resp.json()["access_token"]
        else:
            resp.raise_for_status()

    def _auth_header(self):
        return {"Authorization": "Bearer " + self.access_token}

    def _get_with_retry(self, url: str, params: dict) -> requests.Response:
        if not self._authenticated:
            params["apikey"] = self.app_id
            resp: requests.Response = requests.get(url, params=params)
            if resp.status_code == 200:
                return resp
            else:
                resp.raise_for_status()

        resp: requests.Response = requests.get(
            url, params=params, headers=self._auth_header()
        )

        if resp.status_code == 200:
            return resp
        elif resp.status_code == 401:
            self._update_access_token()
            resp = requests.get(url, params=params, headers=self._auth_header())
            if resp.status_code == 200:
                return resp

        resp.raise_for_status()

    @auth_required
    def _post_with_retry(self, url, data) -> requests.Response:
        resp: requests.Response = requests.post(
            url, json=data, headers=self._auth_header()
        )

        if resp.status_code == 200:
            return resp
        elif resp.status_code == 401:
            self._update_access_token()
            resp = requests.post(url, json=data, headers=self._auth_header())
            if resp.status_code == 200:
                return resp

        # resp will contain the latest http call response
        resp.raise_for_status()

    def quotes(self, symbols: List[str]) -> Dict[str, Quote]:
        params = {"symbol": (",".join(symbols)).upper()}
        resp: requests.Response = self._get_with_retry(Urls.quote, params=params)
        output = resp.json()
        output = {k: Quote(v) for k, v in output.items()}
        if output:
            return output
        else:
            raise SymbolNotFound(f"{','.join(symbols)} not found")

    def quote(self, symbol: str) -> Quote:
        symbol = symbol.upper()
        output = self.quotes([symbol])
        return output[symbol]

    def stock(self, symbol: str) -> Stock:
        output = self.quote(symbol)
        return Stock(output._get_data())

    def find_instrument(self, symbol_pattern: str) -> Dict[str, Instrument]:
        params = {"symbol": symbol_pattern, "projection": "symbol-regex"}
        resp: requests.Response = self._get_with_retry(Urls.search, params=params)
        output = resp.json()
        output = {k: Instrument(v) for k, v in output.items()}
        return output

    def get_fundamentals(self, symbol: str) -> Fundamental:
        symbol = symbol.upper()
        params = {"symbol": symbol, "projection": "fundamental"}
        resp: requests.Response = self._get_with_retry(Urls.search, params=params)
        output = resp.json()
        return Fundamental(output[symbol]["fundamental"])

    def get_history(
        self,
        symbol: str,
        start_dt: datetime = None,
        end_dt: datetime = datetime.today(),
        freq: str = "d",
        outside_rth: bool = False,
    ) -> List[Dict[str, float]]:

        symbol = symbol.upper()
        # TODO Check if date is UTC, if non-tzaware convert appropriately
        if start_dt is None or end_dt is None:
            raise InvalidArgument("Start Date and End Date are required")
        if end_dt < start_dt:
            raise InvalidArgument("Start Date should be before End Date")
        if freq not in ["d", "w", "m", "1min", "5min", "10min", "15min", "30min"]:
            raise InvalidArgument(
                "Frequency should be one of d, w, m, 1min, 5min, 10min, 15min, 30min"
            )
        if "min" in freq and datetime.now() - start_dt > timedelta(days=30):
            raise InvalidArgument(
                "Start Date cannot be more than 30 calendar days ago for intraday data"
            )

        config = {
            "d": ("year", "daily", 1),
            "w": ("year", "weekly", 1),
            "m": ("year", "monthly", 1),
            "1min": ("day", "minute", 1),
            "5min": ("day", "minute", 5),
            "10min": ("day", "minute", 10),
            "15min": ("day", "minute", 15),
            "30min": ("day", "minute", 30),
        }
        periodType, freqType, frequency = config[freq]
        params = {
            "periodType": periodType,
            "frequencyType": freqType,
            "frequency": frequency,
            "needExtendedHoursData": outside_rth,
            "startDate": int(start_dt.timestamp()) * 1000,
            "endDate": int(end_dt.timestamp()) * 1000,
        }
        resp: requests.Response = self._get_with_retry(
            Urls.history % symbol, params=params
        )
        output = resp.json()

        if output["empty"]:
            return None
        else:
            return output["candles"]

    def get_history_df(
        self,
        symbol: str,
        start_dt: datetime = None,
        end_dt: datetime = datetime.today(),
        freq: str = "d",
        outside_rth: bool = False,
    ):
        import pandas as pd

        output = self.get_history(
            symbol, start_dt=start_dt, end_dt=end_dt, freq=freq, outside_rth=outside_rth
        )
        if output is None:
            return None

        df = pd.DataFrame(output)
        df["datetime"] = pd.to_datetime(df["datetime"], unit="ms")

        df.set_index("datetime", inplace=True)
        return df

    def get_expirations(self, symbol: str = None) -> List[str]:
        symbol = symbol.upper()
        params = {
            "symbol": symbol,
            "contractType": "CALL",
            "strategy": "SINGLE",
            "range": "NTM",
        }
        resp: requests.Response = self._get_with_retry(Urls.option_chain, params=params)
        output = resp.json()
        expiries = []
        for v in output["callExpDateMap"].keys():
            d, n = v.split(":")
            expiries.append(d)

        return expiries

    def get_option_chain(self, symbol: str = None, expiry: str = None) -> OptionChain:
        if symbol is None or expiry is None:
            raise InvalidArgument("symbol and expiry (yyyy-mm-dd) are required")
        symbol = symbol.upper()

        params = {
            "symbol": symbol,
            "strategy": "SINGLE",
            "fromDate": expiry,
            "toDate": expiry,
        }
        resp: requests.Response = self._get_with_retry(Urls.option_chain, params=params)
        output = resp.json()
        calls: Dict[str, Option] = {}
        puts: Dict[str, Option] = {}
        for exp, options in output["callExpDateMap"].items():
            if expiry in exp:
                for s in options.keys():
                    calls[s] = Option(options[s][0])
        for exp, options in output["putExpDateMap"].items():
            if expiry in exp:
                for s in options.keys():
                    puts[s] = Option(options[s][0])
        chain = OptionChain(calls, puts)
        return chain

    def get_option(
        self,
        symbol: str = None,
        expiry: str = None,
        right: str = None,
        strike: float = None,
    ) -> Option:
        symbol = symbol.upper()
        if right.lower() in ["c", "call"]:
            right = "CALL"
        if right.lower() in ["p", "put"]:
            right = "PUT"

        ctype_map = {"CALL": "callExpDateMap", "PUT": "putExpDateMap"}
        params = {
            "symbol": symbol,
            "strategy": "SINGLE",
            "fromDate": expiry,
            "toDate": expiry,
            "strike": strike,
            "contractType": right,
        }
        resp: requests.Response = self._get_with_retry(Urls.option_chain, params=params)
        output = resp.json()
        data = output[ctype_map[right]]
        if not data:
            raise SymbolNotFound("Option Not Found")

        out = None
        for e in data.keys():
            if expiry in e:
                out = Option(data[e][Option.float_to_strike(strike)][0])
                return out
